src/Http/Http.Extensions/src/JsonOptions.cs

```diff
@@ -21,11 +23,24 @@ public class JsonOptions
         // Because these options are for producing content that is written directly to the request
         // (and not embedded in an HTML page for example), we can use UnsafeRelaxedJsonEscaping.
         Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
+
+        // The JsonSerializerOptions.GetTypeInfo method is called directly and needs a defined resolver
+        // setting the default resolver (reflection-based) but the user can overwrite it directly or calling
+        // .AddContext<TContext>()
+#pragma warning disable IL2026 // Suppressed in ILLink.Suppressions.LibraryBuild.xml 
+        TypeInfoResolver = CreateDefaultTypeResolver()
+#pragma warning restore IL2026 // Suppressed in ILLink.Suppressions.LibraryBuild.xml
+
     };
 
     // Use a copy so the defaults are not modified.
     /// <summary>
     /// Gets the <see cref="JsonSerializerOptions"/>.
     /// </summary>
     public JsonSerializerOptions SerializerOptions { get; } = new JsonSerializerOptions(DefaultSerializerOptions);
+
+    [RequiresUnreferencedCode("System.Text.Json.Serialization.Metadata.DefaultJsonTypeInfoResolver might require types that cannot be statically analyzed and might need runtime code generation.")]
+    [RequiresDynamicCode("System.Text.Json.Serialization.Metadata.DefaultJsonTypeInfoResolver might require types that cannot be statically analyzed and might need runtime code generation. Enable EnsureJsonTrimmability feature switch for native AOT applications.")]
```

Comment:
Am I correct in understanding that after we have https://github.com/dotnet/sdk/pull/29719, we'll update the `JsonOptions` static constructor to stop calling this method when the "EnsureJsonTrimmability" feature switch is set?

If we're going to mention the feature switch, we'd probably also want to include the full name, "Microsoft.AspNetCore.EnsureJsonTrimmability", and link to https://learn.microsoft.com/dotnet/core/runtime-config/#runtimeconfigjson, but I think picking one of the msbuild properties and telling people to add it their csproj would be simpler.

Setting any one of "EnsureAspNetCoreJsonTrimmability", "PublishTrimmed", "IsTrimmable" or "PublishAot" properties to true should be sufficient, right? Will warning fire for unreferenced or dynamic code when these properties are not set?

I don't think mentioning System.Text.Json.Serialization.Metadata.DefaultJsonTypeInfoResolver is relevant since the user is not directly calling it and probably has no idea why it's getting called. If you remove these attributes, do warnings still propagate from the DefaultJsonTypeInfoResolver? If so, we might as well just leave the original warning, since it's not actionable just yet.

Reply:
> Am I correct in understanding that after we have [dotnet/sdk#29719](https://github.com/dotnet/sdk/pull/29719), we'll update the `JsonOptions` static constructor to stop calling this method when the "EnsureJsonTrimmability" feature switch is set?

Exactly and the private method will be trimmed. Something like:

```c#
 TypeInfoResolver = TrimmingAppContextSwitches.EnsureJsonTrimmability ? null : CreateDefaultTypeResolver()
```

Reply:
> If we're going to mention the feature switch, we'd probably also want to include the full name, "Microsoft.AspNetCore.EnsureJsonTrimmability", and link to https://learn.microsoft.com/dotnet/core/runtime-config/#runtimeconfigjson, but I think picking one of the msbuild properties and telling people to add it their csproj would be simpler.
> 
> Setting any one of "EnsureAspNetCoreJsonTrimmability", "PublishTrimmed", "IsTrimmable" or "PublishAot" properties to true should be sufficient, right? 

Right, the idea is when `EnsureAspNetCoreJsonTrimmability` is set to `true`, that would be possible by setting directly or using the `PublishTrimmed`, `PublishAot` properties, the RuntimeHostConfigurationOption (`Microsoft.AspNetCore.EnsureJsonTrimmability`) will be set and the linker will be able to do the substitutions.

https://github.com/dotnet/aspnetcore/pull/45886/files

> Will warning fire for unreferenced or dynamic code when these properties are not set?

Yes.

Reply:
Also, once you have `EnsureAspNetCoreJsonTrimmability` set to `true` the `TypeInfoResolver` will default `null` when run (F5) or publish, so, developers can understand the behavior during the app development.

Reply:
> I don't think mentioning System.Text.Json.Serialization.Metadata.DefaultJsonTypeInfoResolver is relevant since the user is not directly calling it and probably has no idea why it's getting called. If you remove these attributes, do warnings still propagate from the DefaultJsonTypeInfoResolver? If so, we might as well just leave the original warning, since it's not actionable just yet.

You are right and it will propagate. I will leave the original warning for now. Thanks.